### DISCUSSION TITLE
Is there a way to override the "id" as a default filter in findOne/deleteOne/updateOne endpoint?

### DISCUSSION THREAD

**Discussion by jhonnydelima** [doug-martin/nestjs-query#1506] [Q&A]
I need to create a graphql query that searches for establishments based on their code. With that said, I was thinking about using the query "findOne" which is automatically generated, however, it uses the "id" as the default filter. Is it possible to override this filter with the code (in my case)?

**Comment by smolinari** [comment]
@jhonnydelima - What database are you using? You could theoretically replace the id (its value) with the code you are getting for the establishments and for just that entity, If you can do that, I believe the "one" methods will still work. 

Scott

**Reply by jhonnydelima** [reply]
I'm using **postgres**.

I'm not sure I understood what you said. Is it possible to change only its value? I'd like only the code to appear in the graphql documentation, not the id anymore.

To explain a little bit better what I need to do, the default **read one** method gets "id" as a default filter, like this:
```
{
	establishment(id: ID_NUMBER) {
		id
		code
		...
	}
}
```

And I'd like to override the id with the code, like this:
```
{
	establishment(code: CODE_NUMBER) {
		id
		code
		...
	}
}
```

I was wondering if it's possible to change/add something in "establishments.module" or even in the DTO class.

> establishments.module.ts

```ts
@Module({
  imports: [
    NestjsQueryGraphQLModule.forFeature({
      imports: [NestjsQueryTypeOrmModule.forFeature([Establishment])],
      resolvers: [
        {
          DTOClass: EstablishmentDTO,
          EntityClass: Establishment,
          CreateDTOClass: CreateEstablishmentInput,
          UpdateDTOClass: UpdateEstablishmentInput,
          read: {
            one: {
              -> HERE?
            }
          }
        },
      ],
    }),
  ],
  providers: [],
})
```

Jhonny

**Comment by smolinari** [comment]
So you are using TypeORM or Sequelize? 

At any rate, what I mean is, in your entity, you could define the `id` column with the type your establishment code is. Something like (using TypeORM):

@Entity()
export class Establishment {

    @PrimaryColumn()
    id: string;

}

Since the `id` isn't autogenerated now, you'll be responsible for adding in the `id` value in your create method in your establishment service and there you'll add the establishment's id upon creation.  That should allow you to use the `id` with your code values for the "one" service methods (findOne, updateOne, etc). 

What is the establishment id anyway? 

Scott

**Comment by jhonnydelima** [comment]
TypeORM.

Id is the primary key and code is the registered number of a company (establishment).

The thing is, I just started the project for some tests and I haven't created any service or resolver yet, I'm using only the default methods (autogenerated) for now. My establishments module only contains **the DTOs**, **the entity**, and **the module** (posted in the previous comment).

> establishments module structure
```
|modules
|-establishments
|--dto
|---create-establishment.input.ts
|---establishment.dto.ts
|---update-establishment.input.ts
|--entities
|---establishment.entity.ts
|--establishments.module.ts
```
But based on what you said I believe it's not possible to just override the method, maybe I need to create a new query for that. As I said, I don't want the method to only accept the value of code, but also to change the name of the filter.

Jhonny

**Comment by doug-martin** [comment]
@jhonnydelima I'd probably do something like this 

https://github.com/doug-martin/nestjs-query/blob/master/examples/typeorm/src/todo-item/todo-item.resolver.ts

Your query could look like

```
  @Query(() => EstablishmentDTO)
  async findEstablishmentByCode(@Args() input: FindByCodeInput): Promise<ConnectionType<EstablishmentDTO>> {
    // create your filter for the code
    const filter: Filter<TodoItemDTO> = { code: { eq: input.code }};
    // use the query method but set the limit to 1 as you only expect one record.
    const [ code ] = await this.service.query({ filter, paging: {limit: 1} });
    return code;
  }
```

I hope this helps.

**Reply by jhonnydelima** [reply]
Cool!
But since establishments have a **many-to-many relation** with addresses (with custom properties), and I want to return all establishment data including their addresses, is there any way to relate the 3 tables (**establishments, addresses, establishments_addresses**) to return all establishment data in this query?

Here are the entities mentioned: https://github.com/doug-martin/nestjs-query/discussions/1510
